### PR TITLE
Util

### PR DESCRIPTION
**Pull request by awfulbananas**
it's Vector time

### CODE REVIEW

File: src/main/java/frc/utils/Vector.java [hunk 1]
```diff
@@ -1,401 +1,577 @@
...
+    }
+    return Math.sqrt(n);
+  }
+
+  /**
+   * returns the magnitude of this Vector squared more quickly than mag()
+   *
+   * @return the magnitude of this Vector squared
+   */
+  public double sMag() {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: not a great name. squaredMag

File: src/main/java/frc/utils/Vector.java [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+    }
+    return true;
+  }
+
+  /**
+   * gets the angle from the origin to this Vector
+   *
+   * @return the angle from the origin to this Vector
+   */
+  public double ang() {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: `angle()` is a better name. It's not that long

File: src/main/java/frc/utils/Vector.java [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    for (int i = 0; i < m_vals.length; i++) {
+      if (this.m_vals[i] != p.m_vals[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * gets the angle from the origin to this Vector
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: Oriented counterclockwise. (which is the standard orientation)

File: src/main/java/frc/utils/Vector.java [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+  public Vector clampLength(double maxLength) {
+    if (this.sMag() > maxLength * maxLength) {
+      normalize();
+      mult(maxLength);
+    }
+    return this;
+  }
+
+  /**
+   * clamps the value of a Point between the limits given by minimum and maximum Points such that
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: elementwise clamps.

File: src/main/java/frc/utils/Vector.java [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+  }
+
+  /**
+   * restricts this Point to a maximum length, setting it to that length it it it longer, then
+   * returns itself
+   *
+   * @param maxLength the maximum length to be clamped to
+   * @return this Vector
+   */
+  public Vector clampLength(double maxLength) {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java:438]: Add a min length.

File: src/main/java/frc/utils/Vector.java [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    return new Vector(-Y(), X());
+  }
+
+  /**
+   * rotates this Vector the given number of radians around the origin
+   *
+   * @param rot the number of radians to rotate this Vector around the origin
+   * @return this Vector
+   */
+  public Vector rot(double rot) {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: Call the argument theta.

File: src/main/java/frc/utils/Vector.java [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+
+  /**
+   * divides this vector by another as if they're complex numbers, where the x value is the real
+   * component and the y value is the complex component, throwing an error if either Vector has more
+   * than 2 dimensions
+   *
+   * @param o the Vector to divide by
+   * @return this Vector
+   */
+  public Vector cDiv(Vector o) {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: Remove this.

File: src/main/java/frc/utils/Vector.java [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+
+  /**
+   * multiplies this vector by another as if they're complex numbers, where the x value is the real
+   * component and the y value is the complex component, throwing an error if either Vector has more
+   * than 2 dimensions
+   *
+   * @param o the Vector to multiply by
+   * @return this Vector
+   */
+  public Vector cMult(Vector o) {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: Remove this.

File: src/main/java/frc/utils/Vector.java [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+  /**
+   * applies a 2d linear transformation to this Vector, where the transformed location of (1, 0) is
+   * iHatLoc, and the transformed location of (0, 1) is jHatLoc, and is also equivalent to
+   * multiplying the matrix with iHatLoc and jHatLoc as it's columns by this Vector
+   *
+   * @param iHatLoc the transformed location of the x axis basis vector
+   * @param jHatLoc the transformed location of the y axis basis vector
+   * @return this Vector
+   */
+  public Vector matrixTransform(Vector iHatLoc, Vector jHatLoc) {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java:402]: Technically this is a 2D operation called "matmul", but in any case, not sure if we need so maybe just remove.

File: src/main/java/frc/utils/Vector.java [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+    }
+    return n;
+  }
+
+  /**
+   * returns the X component of this Vector
+   *
+   * @return the X component of this Vector
+   */
+  public double X() {
```
Comment (EmDash00) [inline src/main/java/frc/utils/Vector.java]: Thought about it some more. Just use `x()`. Just keeping with style conventions.

**Review by EmDash00 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/utils/Vector.java (hunk 9 above)
Comment (awfulbananas) [reply src/main/java/frc/utils/Vector.java:402]: linear transformations are general enough that we're likely to have at least some use for it

**Review by EmDash00 — APPROVED** [review]

**Pull request merged**